### PR TITLE
Allow setting scissor rectangle on RenderView

### PR DESCRIPTION
**Pull request by jayrulez**
Add scissor to dynamic states in vulkan backend.

Set  initial scissor in begin pass

Set scissor in WorldRendererShared::setupLightPass to match viewport

### CODE REVIEW

**Comment by jayrulez** [comment]
Don't merge, I need to check some errors.

**Comment by jayrulez** [comment]
Looks good now

File: code/Render/Context/RenderBlock.h [hunk 1]
```diff
@@ -244,6 +244,16 @@ class T_DLLCLASS SetViewportRenderBlock : public RenderBlock
...
 };
 
+/*! Set target scissor block.
+ * \ingroup Render
+ */
+class T_DLLCLASS SetScissorRenderBlock : public RenderBlock
+{
+public:
+	Rectangle scissor;
+	virtual void render(IRenderView* renderView) const override final;
```
Comment (apistol78) [inline code/Render/Context/RenderBlock.h:254]: Blank line between last member and methods.

File: code/Render/Types.h [hunk 2]
```diff
@@ -428,6 +428,36 @@ struct Viewport
...
+	Rectangle() = default;
+	Rectangle(int32_t l, int32_t t, int32_t w, int32_t h)
+		: left(l)
+		, top(t)
+		, width(w)
+		, height(h)
+	{
+	}
+
+	bool operator==(const Rectangle& other) const
```
Comment (apistol78) [inline code/Render/Types.h:447]: Why these operators? Doesn't seem to be necessary and if necessary should be using "= default".

File: code/Render/Types.h [hunk 3]
```diff
@@ -428,6 +428,36 @@ struct Viewport
...
 };
 
+/*! Render scissor rectangle. */
+struct Rectangle
+{
+	int32_t left = 0;
+	int32_t top = 0;
+	int32_t width = 0;
+	int32_t height = 0;
+	Rectangle() = default;
```
Comment (apistol78) [inline code/Render/Types.h:438]: Blank line between last member and methods.

File: code/Render/Vulkan/RenderViewVk.cpp [hunk 4]
```diff
@@ -525,6 +525,20 @@ void RenderViewVk::setViewport(const Viewport& viewport)
...
 
+void RenderViewVk::setScissor(const Rectangle& scissor)
+{
+	T_ASSERT(scissorRect.left >= 0);
+	T_ASSERT(scissorRect.top >= 0);
+	T_ASSERT(scissorRect.width >= 0);
+	T_ASSERT(scissorRect.height >= 0);
+
+	const auto& frame = m_frames[m_currentImageIndex];
+	VkRect2D vkScissor = {};
```
Comment (apistol78) [inline code/Render/Vulkan/RenderViewVk.cpp:536]: I'm trying to move all Vulkan structs into C++20 named initializers.

File: code/Render/Vulkan/RenderViewVk.cpp [hunk 5]
```diff
@@ -525,6 +525,20 @@ void RenderViewVk::setViewport(const Viewport& viewport)
...
+	vkScissor.offset = VkOffset2D { scissor.left, scissor.top };
+	vkScissor.extent = VkExtent2D { static_cast<uint32_t>(scissor.width), static_cast<uint32_t>(scissor.height) };
```
Comment (apistol78) [inline code/Render/Vulkan/RenderViewVk.cpp:538]: Not using static_cast for primitive types since it's too verbose, sticking to C-style casts for those.

File: code/World/Shared/WorldRendererShared.cpp [hunk 6]
```diff
@@ -520,7 +520,17 @@ void WorldRendererShared::setupLightPass(
...
+
+						// Set scissor.
+						auto ssrb = renderContext->alloc< render::SetScissorRenderBlock >();
+						ssrb->scissor = render::Rectangle(
+							svrb->viewport.left,
+							svrb->viewport.top,
+							svrb->viewport.width,
+							svrb->viewport.height
+						);
+						renderContext->draw(ssrb);
```
Comment (apistol78) [inline code/World/Shared/WorldRendererShared.cpp:533]: Why is this necessary?

File: code/Render/Types.h (hunk 2 above)
Comment (jayrulez) [reply code/Render/Types.h]: I'm using it in the RmlUi implementation. I'm caching the last set scissor. When RmlUi wants me to set set the scissor, I check the last set scissor against the new one to avoid unnecessary calls to the vulkan backend.
I updated to use the default implementation, but if you would prefer to not have it here, then I can just do a more verbose check in the RmlUi integration.

File: code/World/Shared/WorldRendererShared.cpp (hunk 6 above)
Comment (jayrulez) [reply code/World/Shared/WorldRendererShared.cpp:533]: I'm not sure if this is necessary tbh. 
Since the vp is being changed here, the scissor and vp will have different sizes.
I assume setting the scissor to the same size as the viewport will preserve existing behaviour.

Comment (jayrulez) [reply code/World/Shared/WorldRendererShared.cpp]: I opted to revert this change.

File: code/Render/Types.h (hunk 2 above)
Comment (jayrulez) [reply code/Render/Types.h]: I’ll remove the equality operator, rebase, and squash the fix up commits. I’m not sure if I remember right, but we may lost the review comments after a force push
